### PR TITLE
Add palindrome linked list to repo

### LINKED ISSUE #110: Add "palindrome linked list" to repo
See https://leetcode.com/problems/palindrome-linked-list/.

### PR DESCRIPTION
**Pull request by SteveStoffer**
My implementation of LeetCode's Palindrome problem for a linked list.

https://leetcode.com/problems/palindrome-linked-list/

Closes #110

### CODE REVIEW

**Comment by SteveStoffer** [comment]
I figured I would add this to try to mess with adding a branch to this repo instead of forking.

File: src/algorithm_practice/Datastructure_Algorithms/SLL/palindrome/palindrome.cpp [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+ *     head->next->next->next
+ *            ^           ^
+ *          slow         fast
+ * 
+ *     head->next->next->next
+ *                  ^          ^
+ *                slow        fast
+ * 
+ * Once we find the middle node, the right half of the list needs 
+ * to be reversed. Since we found the middle node, we can use another
```
Comment (domfarolino) [inline src/algorithm_practice/Datastructure_Algorithms/SLL/palindrome/palindrome.cpp:33]: Let's also add a reference to `./reverseList/reverse.cpp`

File: src/algorithm_practice/Datastructure_Algorithms/SLL/palindrome/palindrome.cpp [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+ * palindrome or false otherwise.
+ * 
+ * A naive approach to this problem would be to try to iterate over
+ * the linked list and store the values into an array. Then compare
+ * them to the linked list. In order to be able to solve this with
+ * O(1) space, we need a better approach.
+ * 
+ * In order to do this, we will need to compare the first element with
+ * the last element one at a time while working towards the middle of 
+ * the linked list. This means we need to reverse half of the linked
```
Comment (domfarolino) [inline src/algorithm_practice/Datastructure_Algorithms/SLL/palindrome/palindrome.cpp:16]: After "middle of the linked list" could you reference `./middleNode/middle.cpp`, just to point the reader to some supplemental documentation on this sub-problem over there?

File: src/algorithm_practice/Datastructure_Algorithms/SLL/palindrome/palindrome.cpp [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+ * pointer to the last value in the list. (We'll call it tail since it
+ * is the last item in the list).
+ *                  
+ *     head->next->next<-next
+ *      ^                  ^
+ *     curr              tail
+ * 
+ * Now we can compare the values while moving the pointers 'towards' 
+ * eachother. If the values aren't the same, return false. If the
+ * pointers reach the end, we can return true!
```
Comment (domfarolino) [inline src/algorithm_practice/Datastructure_Algorithms/SLL/palindrome/palindrome.cpp:50]:
```suggestion
 * pointers reach their respective "ends" of the list, we can return true!
```

File: src/algorithm_practice/Datastructure_Algorithms/SLL/palindrome/palindrome.cpp [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+  
+  printNode(palindrome);
+  std::cout << " is" << (isPalindrome(palindrome) ? " " : " not ") << "a palindrome.\n";
+  
+  deleteList(palindrome);
+
+  ListNode<int>* not_palindrome = new ListNode<int>(1);
+  not_palindrome->next = new ListNode<int>(2);
+  not_palindrome->next->next = new ListNode<int>(2);
+  // not_palindrome->next->next->next = new ListNode<int>(2);
```
Comment (domfarolino) [inline src/algorithm_practice/Datastructure_Algorithms/SLL/palindrome/palindrome.cpp:173]: Is this meant to be uncommented or deleted?

File: src/algorithm_practice/Datastructure_Algorithms/SLL/palindrome/palindrome.cpp [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+ * Time complexity: O(n)
+ * Space complexity: O(1)
+ */
+template <typename T>
+ListNode<T>* reverse(ListNode<T>* head) {
+  if (!head) return nullptr;
+
+  ListNode<T>* prev = nullptr;
+  ListNode<T>* curr = head;
+  ListNode<T>* next;
```
Comment (domfarolino) [inline src/algorithm_practice/Datastructure_Algorithms/SLL/palindrome/palindrome.cpp:80]: Could you initialize this to nullptr (or `head->next`, whatever really) just as a matter of good practice? Uninitialized ptrs are nasty!

File: src/algorithm_practice/Datastructure_Algorithms/SLL/palindrome/palindrome.cpp [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+ * Time complexity: O(n)
+ * Space complexity: O(1)
+ */
+template <typename T>
+bool isPalindrome(ListNode<T>* head) {
+  if (!head) return false;
+
+  ListNode<T>* curr = head;
+  ListNode<T>* tail = reverse(middleNode(head));
+  ListNode<T>* tail_head = tail; // So we can delete the full LL.
```
Comment (domfarolino) [inline src/algorithm_practice/Datastructure_Algorithms/SLL/palindrome/palindrome.cpp:121]: This variable name is just a tad confusing to me haha. How does `original_tail` or `original_tail_node` sound?

**Review by domfarolino — COMMENTED** [review]
Looking good, thanks a lot for taking this one on! Mostly have just a few nits.

File: src/algorithm_practice/Datastructure_Algorithms/SLL/palindrome/palindrome.cpp (hunk 4 above)
Comment (SteveStoffer) [reply src/algorithm_practice/Datastructure_Algorithms/SLL/palindrome/palindrome.cpp:173]: So I had that in there to kind of show that I tested it with an odd amount of elements in the linked list and an even amount of elements. Just to make sure that it still returned false and did not have any undefined behavior. I can uncomment it or remove it if you'd like. Or could I just add another comment to the side saying why I commented it out?

File: src/algorithm_practice/Datastructure_Algorithms/SLL/palindrome/palindrome.cpp (hunk 5 above)
Comment (SteveStoffer) [reply src/algorithm_practice/Datastructure_Algorithms/SLL/palindrome/palindrome.cpp:80]: True! I guess doing it in leet code is one thing, but outside of that I shouldn't do that! I knew better.

File: src/algorithm_practice/Datastructure_Algorithms/SLL/palindrome/palindrome.cpp (hunk 6 above)
Comment (SteveStoffer) [reply src/algorithm_practice/Datastructure_Algorithms/SLL/palindrome/palindrome.cpp:121]: Yes I kind of confused my self on this to try to come up with a good name. But those sound better!

Comment (domfarolino) [reply src/algorithm_practice/Datastructure_Algorithms/SLL/palindrome/palindrome.cpp:121]: Cool, sounds great. I'll take this out of my inbox and just let me know when you've made the changes.

File: src/algorithm_practice/Datastructure_Algorithms/SLL/palindrome/palindrome.cpp (hunk 1 above)
Comment (domfarolino) [reply src/algorithm_practice/Datastructure_Algorithms/SLL/palindrome/palindrome.cpp:33]: I could go either way, whichever you'd like!

**Comment by SteveStoffer** [comment]
Everything should be fixed now. When I referenced the sub problems, I did not add the link since the link was really long. I also changed all of the arrows the the ascii arrows.

**Review by domfarolino — APPROVED** [review]
Great, LGTM! Thanks a lot

**Pull request merged**